elasticdl/python/master/servicer.py

```diff
@@ -33,34 +33,31 @@ def get_task(self, request, _):
         res = elasticdl_pb2.Task()
         res.model_version = self._version
         res.minibatch_size = self._minibatch_size
-        if request.task_type == elasticdl_pb2.EVALUATION:
-            task_id, task = self._task_d.get_eval_task(request.worker_id)
-        else:
-            task_id, task = self._task_d.get(request.worker_id)
 
-        if task:
-            res.task_id = task_id
-            res.shard_name = task.shard_name
-            res.start = task.start
-            res.end = task.end
-            res.type = task.type
-            for k, v in task.extended_config.items():
-                res.extended_config[k] = v
+        while True:
+            if request.task_type == elasticdl_pb2.EVALUATION:
+                task_id, task = self._task_d.get_eval_task(request.worker_id)
+            else:
+                task_id, task = self._task_d.get(request.worker_id)
 
-            # For evaluation task, it will use the fixed version model
-            if task.type == elasticdl_pb2.EVALUATION:
-                res.model_version = task.model_version
-        elif (not self._task_d.finished()) or (
-            self._task_d.invoke_deferred_callback()
-        ):
-            # If the todo and doing tasks are not empty,
-            # Otherwise if the callback list is not empty,
-            # we are trying to pop and invoke the callback.
-            # Then the master tells the worker to wait
-            # in case of new tasks later.
-            res.type = elasticdl_pb2.WAIT
+            if task:
+                res.task_id = task_id
+                res.shard_name = task.shard_name
+                res.start = task.start
+                res.end = task.end
+                res.type = task.type
+                for k, v in task.extended_config.items():
+                    res.extended_config[k] = v
 
-        return res
+                # For evaluation task, it will use the fixed version model
+                if task.type == elasticdl_pb2.EVALUATION:
+                    res.model_version = task.model_version
+                return res
+            elif self._task_d.invoke_deferred_callback():
+                continue
+            else:
+                res.type = elasticdl_pb2.END
```

Comment:
If there are `doing` tasks, the master will still send elasticdl_pb2.END to stop idle workers. They should not stop as when those `doing` tasks may fail and are put back into `todo` due to worker pod preemption or OOM.

Reply:
If the worker pod is killed due to preemption or OOM, there are two things happen:

- the tasks in`doing`  of the worker are pushed `todo`
- the master will relaunch a new worker

The new worker could get work from master and execute it.

Please refer to https://github.com/sql-machine-learning/elasticdl/blob/016a80c513cb1736069f44b004d0e16bdb5e955c/elasticdl/python/master/k8s_instance_manager.py#L247-L253

Reply:
Before this PR:
```
elif (not self._task_d.finished()) or (
            self._task_d.invoke_deferred_callback()
```
So the save model task in deferred_callback will be executed after all training tasks are done.
In this PR, even if there are doing training tasks, as long as there are idle workers, save model task will be called before all training tasks are done.